### PR TITLE
[WIP] Firm Questionnaire

### PR DESCRIPTION
**Pull request by lshepstone**
Firms questionnaire, step 1 - work in progress. Things to consider:

1) Routing declarations, particularly the hacky `only: []`
2) Dedicated controller with only actions for `get` and `post` to questionnaire steps
3) Possibly using Virtus to split out the address part of the questionnaire
4) Additional validations around the postcode, and perhaps for field lengths as well

@benlovell if you have a moment, let me know your thoughts.

### CODE REVIEW

**Comment by alexwllms** [comment]
I've done as much as I can adding dough form helpers to the questionnaire.

Need to review the content with Mark/Kemi, and enquire as to how the interaction works with the "greyed out" input elements.

File: app/views/questionnaires/step_1_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  <section class="l-questionnaire">
+  <%= heading_tag(t('questionnaire.step_one.heading'), level: 1) %>
+  <p><%= t('questionnaire.step_one.introduction') %></p>
+
+  <%= form_for @questionnaire, url: step_1_principal_firm_questionnaire_path, builder: Dough::Forms::Builders::Validation do |f| %>
+    <%= f.validation_summary %>
+    <p><%= t('required') %></p>
+
+    <section>
+    <!-- About your firm -->
```
Comment (benbarnett) [inline app/views/questionnaires/step_1_form.html.erb]: Indentation

File: app/views/questionnaires/step_1_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,209 @@
+<div class="l-constrained">
+  <section class="l-questionnaire">
+  <%= heading_tag(t('questionnaire.step_one.heading'), level: 1) %>
+  <p><%= t('questionnaire.step_one.introduction') %></p>
+
+  <%= form_for @questionnaire, url: step_1_principal_firm_questionnaire_path, builder: Dough::Forms::Builders::Validation do |f| %>
```
Comment (benbarnett) [inline app/views/questionnaires/step_1_form.html.erb]: Should have a class of `form` on the form for completeness

File: app/views/questionnaires/step_1_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        <% end %>
+      <% end %>
+
+    <% end %>
+    </section>
+
+    <hr>
+
+    <section>
+    <%= heading_tag(t('questionnaire.step_one.subheading_two'), level: 2) %>
```
Comment (benbarnett) [inline app/views/questionnaires/step_1_form.html.erb]: Indentation

File: app/views/questionnaires/step_1_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+          <% end %>
+        </div>
+      <% end %>
+    <% end %>
+    </section>
+
+    <hr>
+
+    <section>
+    <%= heading_tag(t('questionnaire.step_one.subheading_three'), level: 2) %>
```
Comment (benbarnett) [inline app/views/questionnaires/step_1_form.html.erb]: Indentation

File: app/controllers/questionnaires_controller.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class QuestionnairesController < ApplicationController
+  def step_1_form
```
Comment (benlovell) [inline app/controllers/questionnaires_controller.rb]: Can we use the word `one` here.

File: app/forms/questionnaire_step1_form.rb [hunk 6]
```diff
@@ -0,0 +1,173 @@
+class QuestionnaireStep1Form
```
Comment (benlovell) [inline app/forms/questionnaire_step1_form.rb]: Can we use the word `one` here.

File: app/forms/questionnaire_step1_form.rb [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  include ActiveModel::Model
+
+  attr_accessor :firm_email_address,
```
Comment (benlovell) [inline app/forms/questionnaire_step1_form.rb]: Drop the `firm_` and replace the `main_office` with just `address` I reckon.

File: app/forms/questionnaire_step1_form.rb [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                :free_initial_meeting,
+                :initial_meeting_duration,
+                :initial_advice_fee_structure,
+                :ongoing_advice_fee_structure,
+                :allow_customers_to_pay_for_advice,
+                :minimum_fixed_one_off_fee_amount
+
+  validates :firm_email_address,
+            presence: true,
+            length: {maximum: 50},
```
Comment (benlovell) [inline app/forms/questionnaire_step1_form.rb]: Spaces in the braces here and elsewhere.

File: app/forms/questionnaire_step1_form.rb [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+  validate :accept_customers_from_list_is_valid
+  validate :advice_in_person_list_is_valid
+  validate :advice_by_other_methods_list_is_valid
+  validate :free_initial_meeting_choice_is_valid
+  validate :initial_meeting_duration_choice_is_valid
+  validate :initial_advice_fee_structure_choice_is_valid
+  validate :ongoing_advice_fee_structure_choice_is_valid
+  validate :allow_customers_to_pay_for_advice_choice_is_valid
+
+  def accept_customers_from_options
```
Comment (benlovell) [inline app/forms/questionnaire_step1_form.rb]: As we discussed, the lookup data here and elsewhere should stay in a form object, once we move most of this logic into the AR model.

File: app/views/questionnaires/step_1_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    <p><%= t('questionnaire.step_one.introduction') %></p>
+
+    <%= form_for @questionnaire, url: step_1_principal_firm_questionnaire_path, html: { class: 'form' }, builder: Dough::Forms::Builders::Validation do |f| %>
+      <%= f.validation_summary %>
+      <p><%= t('required') %></p>
+
+
+      <section>
+        <%= heading_tag(t('questionnaire.step_one.about_your_firm.heading'), level: 2, class: 'heading-small') %>
+        <%= callout_instructional "<p>Blackrock Associates, firm registration number 12345678</p>" %>
```
Comment (alexwllms) [inline app/views/questionnaires/step_1_form.html.erb]: @lshepstone this needs to print out the Firm name and registration number

Comment (lshepstone) [inline app/views/questionnaires/step_1_form.html.erb]: @alexwllms this is part of the work already done by @benlovell, so when this branch is rebased on master, we will have that available to us.

File: app/forms/questionnaire_step1_form.rb [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    @allow_customers_to_pay_for_advice_options ||= [
+        'From their own resources',
+        'From funds to be invested',
+        'Either'
+    ]
+  end
+
+  private
+
+  def accept_customers_from_list_is_valid
```
Comment (lshepstone) [inline app/forms/questionnaire_step1_form.rb]: @benlovell, @tomas-stefano I had originally tried this
```
validates :accept_customers_from,
          presence: true,
          inclusion: {
              in: ->(form) { form.accept_customers_from_options }
          }
```

but the problem with this is that this question allows multiple values, which I can't seem to get the inclusion validator to support. Any ideas? I can though change the other custom validations that only require a single value to this style, which is of course better.

**Comment by alexwllms** [comment]
Should this be closed now we're using firm-questionnaire? @lshepstone

**Comment by lshepstone** [comment]
Not needed anymore.

**Pull request closed without merging**